File: examples/javascript/mega_flow/storage.js
// Class hierarchy — inheritance (extends / super), getters/setters,
// static methods — all preserving taint on the way to the sink.
const { execute } = require("./executor");

class Repository {
  constructor(data) {
    this._data = data;
  }

  // Getter propagates taint out of the private field.
  get cmd() {
    return this._data.cmd;
  }

  set cmd(v) {
    this._data = { ...this._data, cmd: v };
  }

  static wrap(data) {
    return new Repository(data);
  }

  run() {
    const c = this.cmd;
    return execute(c);
  }
}

class AuditedRepository extends Repository {
  constructor(data) {
    super(data);
  }

  run() {
    // super-call preserves taint through the inheritance chain.
    return super.run();
  }
}

async function persist(data) {
  const repo = new AuditedRepository(data);
  return repo.run();
}

module.exports = { persist, Repository, AuditedRepository };

File: examples/ruby/mega_flow/storage.rb
# Class hierarchy — inheritance, mix-in modules, attr accessors,
# singleton/class methods — all preserving taint on the way to the sink.
require_relative 'executor'

module Auditable
  def audit!; self; end
end

class Repository
  include Auditable

  attr_reader :data

  def initialize(data)
    @data = data
  end

  # Getter propagates taint out of the instance variable.
  def cmd
    @data[:cmd]
  end

  def self.wrap(data)
    new(data)
  end

  def run
    audit!
    Executor.execute(cmd)
  end
end

class AuditedRepository < Repository
  def run
    # super preserves taint across the inheritance chain.
    super
  end
end

module Storage
  def self.persist(envelope)
    repo = AuditedRepository.wrap(envelope)
    repo.run
  end
end
